flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceImpl.java

```diff
@@ -39,11 +78,282 @@ public long getCurrentProcessingTime() {
+				ProcessingTimeServiceUtil.getProcessingTimeDelay(timestamp, getCurrentProcessingTime()));
+		}
+
+		timer.bind(timerService.registerTimer(timestamp, timer.getCallback(processingTimeCallbackWrapper.apply(target))));
+
+		return timer;
 	}
 
 	@Override
 	public ScheduledFuture<?> scheduleAtFixedRate(ProcessingTimeCallback callback, long initialDelay, long period) {
-		return timerService.scheduleAtFixedRate(processingTimeCallbackWrapper.apply(callback), initialDelay, period);
+		if (isQuiesced()) {
+			return new NeverCompleteFuture(initialDelay);
+		}
+
+		final TimerScheduledFuture<?> timer = new TimerScheduledFuture<Void>(true, this::removeUndoneTimer);
+		undoneTimers.put(timer, Boolean.TRUE);
+
+		// double check to deal with the following race conditions:
+		// 1. canceling timers from the undone table occurs before putting this timer into the undone table
+		//    (see #cancelTimersNotInExecuting())
+		// 2. using the size of the undone table to determine if all timers have done occurs before putting
+		//    this timer into the undone table (see #tryCompleteTimersDoneFutureIfQuiesced())
+		if (isQuiesced()) {
+			removeUndoneTimer(timer);
+			return new NeverCompleteFuture(initialDelay);
+		}
+
+		timer.bind(
+			timerService.scheduleAtFixedRate(
+				timer.getCallback(processingTimeCallbackWrapper.apply(callback)), initialDelay, period));
+
+		return timer;
+	}
+
+	void quiesce() {
+		status.compareAndSet(STATUS_ALIVE, STATUS_QUIESCED);
+	}
+
+	/**
+	 * This is an idempotent method to allow to repeatedly call.
+	 */
+	CompletableFuture<Void> cancelTimersGracefullyAfterQuiesce() {
+		checkState(status.get() == STATUS_QUIESCED);
+
+		if (!timersDoneFutureAfterQuiescing.isDone()) {
+			if (!cancelTimersNotInExecuting()) {
+				return FutureUtils.completedExceptionally(new CancellationException("Cancel timers failed"));
+			}
+			tryCompleteTimersDoneFutureIfQuiesced();
+		}
+
+		return timersDoneFutureAfterQuiescing;
+	}
+
+	@VisibleForTesting
+	int getNumUndoneTimers() {
+		return undoneTimers.size();
+	}
+
+	private boolean isQuiesced() {
+		return status.get() == STATUS_QUIESCED;
+	}
+
+	private void removeUndoneTimer(TimerScheduledFuture<?> timer) {
+		undoneTimers.remove(timer);
+		tryCompleteTimersDoneFutureIfQuiesced();
+	}
+
+	private void tryCompleteTimersDoneFutureIfQuiesced() {
+		if (isQuiesced() && getNumUndoneTimers() == 0) {
+			timersDoneFutureAfterQuiescing.complete(null);
+		}
+	}
+
+	private boolean cancelTimersNotInExecuting() {
+		// we should cancel the timers in descending timestamp order
+		TimerScheduledFuture<?>[] timers = undoneTimers.keySet().toArray(new TimerScheduledFuture<?>[0]);
+		Arrays.sort(timers, Comparator.reverseOrder());
+		for (TimerScheduledFuture<?> timer : timers) {
+			if (!timer.cancel(false) && !timer.canDoneIfCancellationFailure()) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	static final class TimerScheduledFuture<V> implements ScheduledFuture<V> {
```

Comment:
We may have to wait for a long time until the timer is fired and then `removeUndoneTimer` is called, which is the same as the above comment (https://github.com/apache/flink/pull/10151#discussion_r356000446).

Reply:
If we register callback like this (`whenComplete` or other suitable method) then it gets fired when the `future` is canceled  - i.e. when we quiesce the service.

Reply:
You mean calling `removeUndoneTimer` when cancelling, right? If yes, that is the current implementation and we need to handle the status to make sure that if the timer is in executing when cancelling, the `removeUndoneTimer` must be called after the execution is finished, not in `cancle()`.

For example:
t1 starts execution, state is currently EXECUTING
t2 calls #cancel(interrupt = false)
t2 exit #cancel(interrupt = false)
t1 ends execution and invoking removeUndoneTimer

But another case is:
t1 state is currently PENDING
t2 calls #cancel(interrupt = false)
t2 CASes state from PENDING to CANCELLED
t2 invoking removeUndoneTimer
t2 exit #cancel(interrupt = false)
t1 no longer executes #onProcessingTime

Reply:
I'm a little bit confused now :)

According to `TimerScheduledFuture#cancel()` and its usage, the intended behavior is:
a) prevent any scheduled timers from running
b) cancel them in the underlying `ScheduledExecutorService`
c) throw an exception - **without waiting** - if at least one timer is running after that, then  (`StreamOperatorWrapper#closeGracefully`)

If so, 
1. please update the PR description
2. no need for `timersDoneFutureAfterQuiescing` future
3. to a) prevent from running: just check service status in the timer itself before execution
4. to c) throw an Exception: add `volatile Timer#isRunning`, set/unset while running, if any timer **after** quiescing is running then throw an exception

(to cancel timers (b) we still need to store them in some collection)

Reply:
Not.  It may be that my previous expression is not accurate. The intended behavior of `TimerScheduledFuture#cancel(interrupt = false)` is:

For one shot timer:
1. If  `#onProcessingTime()` has not been executed, prevents it from executing. Then triggers  
 `#removeUndoneTimer()` and returns true.
2. Else if `#onProcessingTime()`  is executing, returns false. `onProcessingTime()` will continue to 
 execute asynchronously and trigger `#removeUndoneTimer()` after execution is finished.
3. Else returns false. (`#onProcessingTime()`  has already completed normally or abnormally)

For periodic timer:
1. If `#onProcessingTime()` is not in execution, prevents it from executing and ensure that `#onProcessingTime()` will not be fired to execute in the future. Then triggers `#removeUndoneTimer()` and returns true.
2. If `#onProcessingTime()` is executing, try to call the underlying `ScheduledFuture#cancel(interrupt = false)` to ensure that `#onProcessingTime()` will not be fired to execute in the future. Returns true if successful, false otherwise. The currently executing ``onProcessingTime()` will continue to execute asynchronously and trigger #removeUndoneTimer()`  after execution is finished..

I'll comments on "timerscheduledfuture cancel()"
The current "timerscheduledfuture cancel()" is missing comments and I will add them.

Reply:
About its usage in `ProcessingTimeServiceImpl#cancelTimersNotInExecuting()`, false will be returned only when the timer be cancelled failed and its `#onProcessingTime()` has not been scheduled for execution.

Reply:
Summarizing the offline discussion:
1. PR desc to be updated
2. `status` field is needed for future (2nd PR which involves API and state changes). Currently, a simple solution will suffice:
```
volatile bool quiesced = false;
AtomicInt cnt = 0;

in RunTimer:
  if (quiesced)  return;
  cnt++;
  try {
    if (!quiesced)  { // double check to avoid race condition (comleting future and still running this timer)
    ... timer logic ...
    }
  } finally {
    cnt--;
    if (cnt == 0 &&  quiesced) service.future.complete();
  }

in quiesce():
  quiesced = true;
  if (cnt == 0) service.future.complete();
```

That is:
1. we don't allow timers to fire if quiesced
2. if no timers running, complete future now
3. if any timer is running, it will complete future when finished